scripts/train_cz_gate.py

```diff
@@ -0,0 +1,273 @@
+#
+# Copyright 2021-2024 Budapest Quantum Computing Group
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import piquasso as pq
+
+import tensorflow as tf
+import numpy as np
+import time
+
+from piquasso.decompositions.clements import (
+    get_weights_from_interferometer,
+    get_interferometer_from_weights,
+)
+from piquasso._math.indices import get_operator_index
+
+from typing import Callable
+
+
+LR = 0.00025
+ITERATIONS = 1000
+ALPHA = 10
+BETA = 10_000
+S_STAR = 0.075
+
+tf.get_logger().setLevel("ERROR")
+
+
+def get_expected_density_matrix(
+    state_vector: tf.Tensor, cutoff: int, calculator: pq.TensorflowCalculator
+) -> tf.Tensor:
+    state_00 = [0, 1, 0, 1]
+    state_01 = [0, 1, 1, 0]
+    state_10 = [1, 0, 0, 1]
+    state_11 = [1, 0, 1, 0]
+
+    config = pq.Config(normalize=False, cutoff=cutoff)
+    expected_program = pq.Program(
+        instructions=[
+            pq.StateVector(state_00, coefficient=state_vector[0]),
+            pq.StateVector(state_01, coefficient=state_vector[1]),
+            pq.StateVector(state_10, coefficient=state_vector[2]),
+            pq.StateVector(state_11, coefficient=-state_vector[3]),
+        ]
+    )
+
+    simulator = pq.PureFockSimulator(d=4, config=config, calculator=calculator)
+    expected_state = simulator.execute(expected_program).state
+
+    return expected_state.density_matrix
+
+
+def get_ideal_weights():
```

Comment:
`get_initial_weights` maybe? This could be confusing because these weights are not ideal when imperfect detectors are used.